### PR TITLE
tools: add quant_ablation.py — A/B any quantization scheme against fp16, engine-free

### PR DESCRIPTION
**Pull request by dnnspaul**
Requested in #108. An engine-free way to A/B a quantization scheme against fp16 **before** committing to a 370 GB conversion.

## What it does

Comparing colibrì's score to a published model-card number can't measure quantization: this harness scores 0-shot log-likelihood, published numbers are few-shot/CoT, and that protocol gap can swamp the effect entirely.

This removes the confound **by construction** — take an fp16 model, push its weights through colibrì's *own* quantizer (quantize → dequantize, in place), and score both with the **same harness, same questions, same machine**. The only variable is the quantizer, so the delta *is* the quantization cost.

Quantizer math is replicated from `tools/convert_fp8_to_int4.py` (symmetric absmax, per-row scales) and generalised with an optional group size, so grouped schemes can be compared head-to-head with what ships today. Scheme grammar: `fp16 | int{2,4,8}[-g<N>][-nohead]`.

```sh
python tools/fetch_benchmarks.py --out ./bench --tasks hellaswag,arc_challenge,mmlu --limit 200
python tools/quant_ablation.py --model allenai/OLMoE-1B-7B-0924 --data ./bench \
    --tasks hellaswag,arc_challenge,mmlu --limit 200 \
    --schemes fp16,int4,int4-g64,int4-g128
```

## Results it produced (OLMoE-1B-7B, n=200/task — #108)

| scheme | hellaswag | arc_c | mmlu | mean | Δ vs fp16 |
|---|---|---|---|---|---|
| fp16 | 77.0% | 47.0% | 47.0% | 57.0% | — |
| **int4 (as shipped)** | 74.0% | 41.0% | **31.5%** | 48.8% | **−8.2pp** |
| int4-nohead | 73.5% | 40.5% | 37.5% | 50.5% | −6.5pp |
| **int4-g128** | 78.5% | 45.5% | 38.0% | 54.0% | **−3.0pp** |
| int4-g128-nohead | 78.5% | 46.5% | 38.0% | 54.3% | −2.7pp |

The per-row int4 container costs ~8pp, **concentrated on the hard task** — MMLU falls to 31.5% against a 25% random baseline while easy HellaSwag barely moves. Per-row scales eat the small logit margins hard questions depend on: the same margin erosion that flips near-tie tokens in #100. `group=128` recovers ~63% of it. Keeping `lm_head`/`embed` in fp16 is **not** the fix (+1.7pp alone, +0.3pp on top of grouping).

## Coverage assert (please keep this)

`transformers` fuses MoE experts into 3D tensors (`mlp.experts.gate_up_proj = [E, in, out]`) even when the checkpoint stores one 2D matrix per expert. A `p.ndim == 2` filter therefore skips **every expert**, leaving ~85% of the model in fp16 while appearing to work — my first run did exactly that and produced a plausible, entirely fictional number. The tool prints quantized-param coverage and **exits non-zero below `--min-coverage` (default 95%)**, so that failure is loud instead of believable.

## Notes

- Dev-only tool (`torch` + `transformers`); the engine's dependency-free path is untouched.
- Scoring mirrors `tools/eval_glm.py` exactly, including `acc_norm` normalising by **character** length of the continuation (not tokens) and the same seed/shuffle.
- `make check` passes (38 tests, clean build).
- Targeted at `dev` per CONTRIBUTING.md.

Should slot straight into ranking grouped-scale / error-feedback / rotation (#81) schemes in minutes rather than days.

### CODE REVIEW

**Review by JustVugg — APPROVED** [review]
Exactly the tool the quantization work needs — thank you for turning the offer into code. Security read: clean (the only `eval` is `model.eval()`, PyTorch mode-setting; no os.system/subprocess/shell/pickle/raw-network), standard ML deps (torch/transformers), and it's a self-contained research script in `tools/` — zero engine impact, can't affect default users. It does the confound-free thing you proved in #108: fake-quant (quantize→dequantize→fp16 matmul) to isolate weight error, apply-scheme parametrized (grouped, per-row, head/embed exclusions), and it should include the ~100% coverage assert to avoid the 3D-fused-experts trap you flagged. This lets us rank grouped/error-feedback/rotation (#81) schemes in minutes before a 370 GB conversion — invaluable. Merging to dev.
